docs/architecture/third-party/drupal/testing.md

```diff
@@ -0,0 +1,87 @@
+# Testing Drupal components
+
+OpenEuropa requires automated tests to be written for every new feature or
+bugfix to ensure that the functionality will keep working as expected in the
+future.
+
+* Pull requests without sufficient test coverage will not be accepted.
+* For a pull request to be accepted, both the previously existing and the newly
+  added tests should be passing.
+* Existing test code should not be modified unless this is covered by a change
+  request originating from project stakeholders.
+
+## Choosing the right test framework
+
+In addition to the built-in testing frameworks that ship with Drupal core,
+OpenEuropa also uses Behat to describe business requirements. This section
+details which test framework to use under which circumstances.
+
+### Behat
+
+Whenever a PR intends to change end user behaviour it is required to describe
+the expected user behaviour in a Behat scenario. In practice this means:
+
+* Every user story is accompanied by a Behat scenario. This scenario will be
+  provided to the project stakeholders for acceptance testing.
+* The target audience for these scenarios are the project stakeholders. The
+  Behat scenarios prove to project stakeholders that the project fulfills all
+  business requirements.
+* Every Behat test scenario is written in domain specific business language.
+  The scenarios should avoid the use of technical terms, machine names or
+  references to underlying technology which might not be understood by
+  non-technical stakeholders.
+* Behat tests should only be used to describe expected user behaviour as
+  specified by project stakeholders. It should *not* be used to test code. Any
+  code written that does not directly affect the expected end user behaviour
+  should be covered by unit tests instead. At no point the business should see
+  a Behat scenario that describes a behaviour that was not directly requested
+  by them.
+* For many user stories the Behat scenario will be sufficient to cover the full
+  extent of the code that was written. However, business requirements are known
```

Comment:
Why mentioning the code in this behat (functional) section?
Should the functional test bother to cover most crucial parts of the functionality or most crucial parts of the code added?
or both?

Reply:
It shouldn't cover only the most crucial parts, I have not mentioned this anywhere. It should cover every possible scenario. We should not leave any expected user behaviour untested because then we cannot guarantee to the business that their requested functionality works as expected.